### ISSUE TITLE
Manually edited :GPTEL_SYSTEM: gets overwritten upon save

### ISSUE THREAD

**Issue by nordlow** [karthink/gptel#441]
An edited value of the property `:GPTEL_SYSTEM:` in an org-mode buffer gets overwritten upon save.

For instance if I modify

version 1

```
You are a large language model and a careful programmer. Provide code and only code as output without any additional text, prompt or note.
```

to

version 2

```
You are a large language model and a careful programmer. Provide code and only code as output without any additional text, prompt or note. Write C using as modern constructs as possible.
```

it gets reverted to version 1 which is not what I want. A mode should _never_ overwrite  a users setting by default.

**Comment by karthink** [comment]
Are you setting the `GPTEL_SYSTEM` property (or others) manually?  This is not supported.  gptel saves the active model parameters to the buffer for persistence.  The idea is that the current system message, max token count etc will be written to the file so you can resume the chat with these settings.

> A mode should never overwrite a users setting by default.

`GPTEL_*` properties are not intended to be user settings, they're internal to gptel and used to rehydrate chat sessions.

**Comment by nordlow** [comment]
> Are you setting the `GPTEL_SYSTEM` property (or others) manually? This is not supported. gptel saves the active model parameters to the buffer for persistence. The idea is that the current system message, max token count etc will be written to the file so you can resume the chat with these settings.
> 
> > A mode should never overwrite a users setting by default.
> 
> `GPTEL_*` properties are not intended to be user settings, they're internal to gptel and used to rehydrate chat sessions.

Ok, thanks. So how do I adjust the 

```
:GPTEL_SYSTEM: You are a large language model and a careful programmer. Provide code and only code as output without any additional text, prompt or note.
```

to include adjustments for a specific project? Such as a project containing C code I want to have be generated in a certain manner.

**Comment by nordlow** [comment]
> Are you setting the GPTEL_SYSTEM property (or others) manually?

Yes, I want to change it manually for every Org buffer so that I get a specific prompt for each buffer I'm currenly working in. Ideally I would like to be able to make the prompt customization as fine-grained as possible even on the symbo-level.

**Comment by karthink** [comment]
> So how do I adjust the `GPTEL_SYSTEM` ...

Set the system message for the Org buffer -- you can do it from the transient menu (`C-u C-c RET` in your setup):

![image](https://github.com/user-attachments/assets/ea20bad5-d28e-4199-b2ee-4c366f40fd1c)

You can set it per buffer by toggling the "per-buffer" switch:

![image](https://github.com/user-attachments/assets/ee4d9ad6-083e-4737-892f-998d707d84e4)

> Ideally I would like to be able to make the contextual-prompt-customization as fine-grained as possible even on the symbol-level.

I don't know what it means to set a system message at the symbol level.  But you can set a different system message per Org heading by calling `M-x gptel-org-set-properties`.  See the README for more information on using gptel in Org mode.

----

The active system message is written as the `GPTEL_SYSTEM` property to the Org heading/file so that it can be used the next time the file is opened.  So the intended flow of information is "currently active system message" --> "Org file", _not_ "Org file" --> "active system message".

**Comment by nordlow** [comment]
Ok, thanks. Where is the global system message saved? I've pressed C-x C-s in the transient buffer but I don't get any changes under `~/.emacs.d/transient``.

**Comment by karthink** [comment]
> Ok, thanks. Where is the global system message persistently saved to disk?

It's saved in the Org file, along with other model parameters.

If you would like to have a bunch of pre-written system messages available to quickly select, you should customize `gptel-directives`.  (Please run `M-x customize-group` `gptel` and take a look at the available options)

----

I think there's some miscommunication about what you're trying to do and what I understand you want, because there should be no reason to look in the transient values file for anything.

**Comment by nordlow** [comment]
Ok, thanks. I still strongly believe that manual editing of the `:GPTEL_SYSTEM: ` should be allowed respected upon  of save which then should reload the system prompt. This is behaviour is very similar to how manual editing of calls to existing calls to `custom-set-variables` and `custom-set-faces` in .emacs.d/init.el is respected.

**Comment by karthink** [comment]
> I still strongly believe that manual editings of the `:GPTEL_SYSTEM:
> ` in buffer `b` should be respected upon save of `b`.

I don't see how this can work -- if `GPTEL_SYSTEM` was intended to be edited manually, how would gptel be able to save the active system message to it?  It would become non-functional as a mechanism to save the chat state.

**Comment by nordlow** [comment]
> how would gptel be able to save the active system message to it?
Can you elaborate on what you mean by active system message?

>  if GPTEL_SYSTEM was intended to be edited manually, how would gptel be able to save the active system message to it
It should work both ways just like M-x `customize-variable` picks up the current by parsing the call to `custom-set-variables` and writes it back to init.el upon C-x C-s in the `*Customize ...` buffer.

**Comment by karthink** [comment]
gptel's request parameters aren't like the customize system, for an important reason.  The values in the customize buffer are "dead" until they're saved, at which point they become active.  gptel's parameters stored as Org properties are "live", always active.

I think I didn't explain the problem well.  When you save the chat buffer to disk, there are two sources of truth for the request parameters:

1. The `GPTEL_*` Org properties
2. The currently chosen values for the parameters, for example via `gptel-max-tokens`, `gptel-temperature` etc.

When these are in conflict, there is a problem.  Either source 1 has to override source 2, in which case saving the file does _nothing_.

Or 2 has to override 1, in which case editing the properties by hand will be overwritten.

gptel does 2 because it's the more consistent and automatic approach -- this way you can set whatever properties you want from the menu or via elisp, and just save the file to persist them to disk.  In the other case, the _only_ way to store the request parameters to file would be to edit the properties manually.
